lib/rubocop/config_obsoletion.rb

```diff
@@ -4,298 +4,102 @@ module RuboCop
-      }
-    ].freeze
-
-    attr_reader :warnings
+    # Can be extended by extension libraries to add their own obsoletions
+    self.files = [DEFAULT_RULES_FILE]
 
     def initialize(config)
       @config = config
+      @rules = load_rules
       @warnings = []
     end
 
-    def reject_obsolete_cops_and_parameters
-      messages = [obsolete_cops, obsolete_parameters,
-                  obsolete_enforced_style].flatten.compact
+    def reject_obsolete!
+      messages = obsoletions.flatten.compact
       return if messages.empty?
 
       raise ValidationError, messages.join("\n")
     end
 
     private
 
-    def obsolete_cops
-      OBSOLETE_COPS.map do |cop_name, message|
-        next unless @config.key?(cop_name) ||
-                    @config.key?(Cop::Badge.parse(cop_name).cop_name)
-
-        message + "\n(obsolete configuration found in " \
-                  "#{smart_loaded_path}, please update it)"
-      end
-    end
-
-    def obsolete_enforced_style
-      OBSOLETE_ENFORCED_STYLES.map do |params|
-        obsolete_enforced_style_message(params[:cop], params[:parameter],
-                                        params[:enforced_style],
-                                        params[:alternative])
+    # Default rules for obsoletions are in config/obsoletion.yml
+    # Additional rules files can be added with `RuboCop::ConfigObsoletion.files << filename`
+    def load_rules
+      rules = self.class.files.each_with_object({}) do |filename, hash|
+        hash.merge!(YAML.safe_load(File.read(filename))) do |_key, first, second|
+          first.merge(second)
+        end
       end
-    end
 
-    def obsolete_enforced_style_message(cop, param, enforced_style, alternative)
-      style = @config[cop]&.detect { |key, _| key.start_with?(param) }
+      cop_rules = rules.slice(*COP_RULE_CLASSES.keys)
+      parameter_rules = rules.slice(*PARAMETER_RULE_CLASSES.keys)
 
-      return unless style && style[1] == enforced_style
-
-      "obsolete `#{param}: #{enforced_style}` (for #{cop}) found in " \
-      "#{smart_loaded_path}\n#{alternative}"
+      load_cop_rules(cop_rules).concat(load_parameter_rules(parameter_rules))
     end
 
-    def obsolete_parameters
-      OBSOLETE_PARAMETERS.collect do |params|
-        messages = obsolete_parameter_message(params[:cops], params[:parameters],
-                                              params[:alternative])
-
-        # Warnings are collected separately and not added to the error message
-        if messages && params.fetch(:severity, :error) == :warning
-          @warnings.concat(messages)
-          next
+    # Cop rules are keyed by the name of the original cop
+    def load_cop_rules(rules)
+      rules.each_with_object([]) do |(rule_type, data), out|
+        objects = data.map do |configuration|
+          COP_RULE_CLASSES[rule_type].new(@config, *configuration)
         end
 
-        messages
+        out.concat(objects.compact)
```

Comment:
Cosmetic: it's just a matter of a personal preference, and leaving this completely up to you.
```ruby
rules.flat_map do |rule_type, data|
  data.map do |configuration|
    COP_RULE_CLASSES[rule_type].new(@config, *configuration)
  end
end
```

or
```ruby
rules.flat_map do |rule_type, data|
  rule_class = COP_RULE_CLASSES[rule_type]
  data.map do |configuration|
    rule_class.new(@config, *configuration)
  end
end
```

Reply:
Love it, thanks!